vaadin-upload-flow-parent/vaadin-upload-flow/src/main/java/com/vaadin/flow/component/upload/Upload.java

```diff
@@ -560,17 +561,48 @@ public void setI18n(UploadI18N i18n) {
 
         runBeforeClientResponse(ui -> {
             if (i18n == this.i18n) {
-                JsonObject i18nObject = (JsonObject) JsonSerializer
-                        .toJson(this.i18n);
-                for (String key : i18nObject.keys()) {
-                    ui.getPage().executeJavaScript(
-                            "$0.set('i18n." + key + "', $1)", getElement(),
-                            i18nObject.get(key));
-                }
+                setI18nWithJS();
             }
         });
     }
 
+    private void setI18nWithJS() {
+        JsonObject i18nJson = (JsonObject) JsonSerializer.toJson(this.i18n);
+
+        // Remove null values so that we don't overwrite existing WC
+        // translations with empty ones
+        deeplyRemoveNullValuesFromJsonObject(i18nJson);
+
+        // Assign new I18N object to WC, by deeply merging the existing
+        // WC I18N, and the values from the new UploadI18N instance,
+        // into an empty object
+        getElement().executeJs(
+                "const dropFiles = Object.assign({}, this.i18n.dropFiles, $0.dropFiles);"
```

Comment:
Would it make more sense to use a StringBuilder since a lot of concatenation is happening here?

Reply:
Looks like the compiler will [use StringBuilder under the hood](https://stackoverflow.com/a/1532499/11368281) if it thinks that it would optimize the performance. In terms of readability I prefer the String concatenation to chained `.append`  calls